examples/purchase.rs

```diff
@@ -0,0 +1,145 @@
+enum Command {
+    #[command(description = "display this text.")]
+    Help,
+    #[command(description = "start the purchase procedure.")]
+    Start,
+}
+
+#[tokio::main]
+async fn main() {
+    pretty_env_logger::init();
+    log::info!("Starting dialogue_bot...");
+
+    let bot = Bot::from_env().auto_send();
+
+    Dispatcher::builder(
+        bot,
+        dialogue::enter::<Update, InMemStorage<State>, State, _>()
+            .branch(
+                Update::filter_message()
+                    .branch(teloxide::handler![State::ReceiveFullName].endpoint(receive_full_name))
+                    .branch(dptree::entry().filter_command::<Command>().endpoint(handle_command))
+                    .branch(dptree::endpoint(invalid_state)),
+            )
+            .branch(
+                Update::filter_callback_query().chain(
+                    teloxide::handler![State::ReceiveProductChoice { full_name }]
+                        .endpoint(receive_product_selection),
+                ),
+            ),
+    )
+    .dependencies(dptree::deps![InMemStorage::<State>::new()])
+    .build()
+    .setup_ctrlc_handler()
+    .dispatch()
+    .await;
+}
+
+async fn handle_command(
+    bot: AutoSend<Bot>,
+    msg: Message,
+    cmd: Command,
+    dialogue: MyDialogue,
+) -> HandlerResult {
+    match cmd {
+        Command::Help => {
+            bot.send_message(msg.chat.id, Command::descriptions().to_string()).await?;
+        }
+        Command::Start => {
+            bot.send_message(msg.chat.id, "Let's start! What's your full name?").await?;
+            dialogue.update(State::ReceiveFullName).await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn receive_full_name(
+    bot: AutoSend<Bot>,
+    msg: Message,
+    dialogue: MyDialogue,
+) -> HandlerResult {
+    match msg.text().map(ToOwned::to_owned) {
+        Some(full_name) => {
+            let products = InlineKeyboardMarkup::default().append_row(
+                vec!["Apple", "Banana", "Orange", "Potato"].into_iter().map(|product| {
+                    InlineKeyboardButton::callback(product.to_owned(), product.to_owned())
+                }),
+            );
+
+            bot.send_message(msg.chat.id, "Select a product:").reply_markup(products).await?;
+            dialogue.update(State::ReceiveProductChoice { full_name }).await?;
+        }
+        None => {
+            bot.send_message(msg.chat.id, "Please, send me your full name.").await?;
+        }
+    }
+
+    Ok(())
+}
+
+async fn receive_product_selection(
+    bot: AutoSend<Bot>,
+    q: CallbackQuery,
+    dialogue: MyDialogue,
+    full_name: String,
+) -> HandlerResult {
+    if let Some(product) = &q.data {
+        if let Some(chat_id) = q.chat_id() {
+            bot.send_message(
+                chat_id,
+                format!("{full_name}, product '{product}' has been purchased successfully!"),
+            )
+            .await?;
+        }
+
+        dialogue.update(State::Start).await?;
```

Comment:
shouldn't we only update the state if the operation is successful (we know the `chat_id`)?

Reply:
Yes, and I replaced `.update(...)` with `.exit()`.